File: app/src/main/java/com/example/ceroxlol/remindme/activities/MainActivity.java
package com.example.ceroxlol.remindme.activities;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.viewpager.widget.ViewPager;

import com.example.ceroxlol.remindme.fragments.AppointmentsFragment;
import com.example.ceroxlol.remindme.models.AppointmentKT;
import com.example.ceroxlol.remindme.utils.AppointmentMetCheckingService;
import com.example.ceroxlol.remindme.utils.GpsTracker;
import com.example.ceroxlol.remindme.R;
import com.google.android.material.tabs.TabLayout;

import com.example.ceroxlol.remindme.adapters.MainPageAdapter;

import java.util.Objects;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";
    private final boolean APPOINTMENT_TRACKER_ENABLED = false;

    private GpsTracker gpsTracker;

    private AppointmentMetCheckingService appointmentMetCheckingService;

    private Button buttonAddNewAppointment;
    private Button buttonEditAppointment;
    private Button addNewLocation;
    private Button editLocation;

    private final int REQUEST_APP_PERMISSIONS = 1;
    private final int REQUEST_NEW_FAVORITE_LOCATION = 10;
    private final int REQUEST_EDIT_FAVORITE_LOCATION = 11;
    private final int REQUEST_CREATE_NEW_APPOINTMENT = 20;
    private final int REQUEST_EDIT_APPOINTMENT = 21;

    private MainPageAdapter mainPageAdapter;

    private ActivityResultLauncher<Intent> addAppointmentActivityResultLauncher;

    private NavController navController;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //TODO: Implement good layout first, then: Filter appointments

        init();
    }

    private void init() {
        checkPermissions();
        initClasses();
        initUI();
    }

    private void checkPermissions() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.ACCESS_FINE_LOCATION},
                    REQUEST_APP_PERMISSIONS);
        }
    }

    private void initClasses() {
        NavHostFragment navHostFragment = (NavHostFragment) getSupportFragmentManager().findFragmentById(R.id.nav_host_fragment);
        navController = Objects.requireNonNull(navHostFragment).getNavController();

        gpsTracker = new GpsTracker(getApplicationContext());

        if (APPOINTMENT_TRACKER_ENABLED) {
            appointmentMetCheckingService = new AppointmentMetCheckingService(gpsTracker, this);
            appointmentMetCheckingService.setRun(true);
            appointmentMetCheckingService.start();
        }

        //TODO: Improve the layout for appointments to be shown
        //Make them expandable. Show only name and time?
        buttonAddNewAppointment = findViewById(R.id.buttonAddNewAppointment);
        buttonEditAppointment = findViewById(R.id.buttonEditAppointment);
        addNewLocation = findViewById(R.id.buttonAddNewLocation);
        editLocation = findViewById(R.id.buttonEditLocation);

        ViewPager viewPager = findViewById(R.id.viewPager);
        mainPageAdapter = new MainPageAdapter(getSupportFragmentManager());
        viewPager.setAdapter(mainPageAdapter);

        TabLayout tabLayout = findViewById(R.id.tabLayout);
        tabLayout.setupWithViewPager(viewPager);

        addAppointmentActivityResultLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), this::updateRecyclerview);
    }

    private void updateRecyclerview(ActivityResult result) {
        if (result.getResultCode() == Activity.RESULT_OK) {
            AppointmentKT appointment = Objects.requireNonNull(result.getData()).getParcelableExtra("appointment");
            AppointmentsFragment appointmentsFragment = (AppointmentsFragment) mainPageAdapter.getItem(0);
            if (appointment != null) {
                appointmentsFragment.insertData(appointment);
            }
        }
    }

    private void initUI() {
        this.buttonAddNewAppointment.setOnClickListener(v -> {
            if (!checkIfLocationsAreAvailable()) {
                showNoLocationsAvailableAlertDialog();
            } else {
                Intent i = new Intent(MainActivity.this, AddNewAppointmentActivity.class);
                addAppointmentActivityResultLauncher.launch(i);
            }
        });

        this.buttonEditAppointment.setOnClickListener(v -> {
            Intent i = new Intent(MainActivity.this, EditAppointmentsFragment.class);
            startActivityForResult(i, REQUEST_EDIT_APPOINTMENT);
        });

        this.addNewLocation.setOnClickListener(v -> {
            Intent i = new Intent(MainActivity.this, ChooseLocationActivity.class);
            startActivityForResult(i, REQUEST_NEW_FAVORITE_LOCATION);
        });

        this.editLocation.setOnClickListener(v -> {
            Intent i = new Intent(MainActivity.this, EditLocationActivity.class);
            startActivityForResult(i, REQUEST_EDIT_FAVORITE_LOCATION);
        });
    }

    private void showNoLocationsAvailableAlertDialog() {
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(this);

        // Setting Dialog Title
        alertDialog.setTitle("No locations available");

        // Setting Dialog Message
        alertDialog.setMessage("In order to create appointments you need to have locations first. Please create one.");

        // Setting Icon to Dialog
        //alertDialog.setIcon(R.drawable.delete);

        // On pressing OK button
        alertDialog.setPositiveButton("OK", null);

        // Showing Alert Message
        alertDialog.show();
    }

    private boolean checkIfLocationsAreAvailable() {
        return getDb().locationMarkerDao().entriesExist();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        Log.d(TAG, "requestCode: " + requestCode + "  resultCode: " + resultCode);

        if (requestCode == REQUEST_CREATE_NEW_APPOINTMENT) {
            if (resultCode == RESULT_OK) {
                Log.i(TAG, "Successfully created new appointment.");

            }
        }

        if (requestCode == REQUEST_NEW_FAVORITE_LOCATION) {
            //Do something based on a new location
        }
    }
}

File: app/src/main/java/com/example/ceroxlol/remindme/activities/MainActivity.kt
/*
 * Copyright (C) 2021 The Android Open Source Project.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.ceroxlol.remindme.activities

import android.Manifest
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.provider.Settings
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.MenuProvider
import androidx.navigation.NavController
import androidx.navigation.NavOptions
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI.setupActionBarWithNavController
import androidx.navigation.ui.navigateUp
import androidx.preference.PreferenceManager
import com.example.ceroxlol.remindme.R
import com.example.ceroxlol.remindme.services.LocationService


class MainActivity : AppCompatActivity(R.layout.activity_main) {

    private lateinit var navController: NavController

    private var locationService: LocationService? = null

    // Tracks the bound state of the service.
    private var isServiceBound = false

    // Monitors the state of the connection to the service.
    private val locationServiceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            val binder: LocationService.LocalBinder =
                service as LocationService.LocalBinder
            locationService = binder.service
            isServiceBound = true
            locationService!!.requestLocationUpdates()
        }

        override fun onServiceDisconnected(name: ComponentName) {
            locationService = null
            isServiceBound = false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val navHostFragment = supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as NavHostFragment
        navController = navHostFragment.navController
        val appBarConfiguration = AppBarConfiguration(navController.graph)
        setSupportActionBar(findViewById(R.id.toolbar))
        setupActionBarWithNavController(
            this,
            navController = navController,
            configuration = appBarConfiguration
        )

        addMenuProvider(object : MenuProvider{

            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                val inflater: MenuInflater = menuInflater
                inflater.inflate(R.menu.main_activity_menu, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                // Handle the menu selection
                when (menuItem.itemId) {
                    R.id.action_settings -> {
                        // User chose the "Settings" item, show the app settings UI...
                        val options = NavOptions.Builder()
                            .setEnterAnim(android.R.anim.fade_in)
                            .setEnterAnim(android.R.anim.fade_out)
                            .setPopEnterAnim(android.R.anim.slide_in_left)
                            .setPopExitAnim(android.R.anim.slide_out_right)
                            .build()
                        navController.navigate(R.id.settingsFragment, null, options)
                        return true
                    }
                }
                return false
            }

        }
        )

        checkLocationPermission()
    }

    private fun checkLocationPermission() {
        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(
                    this,
                    Manifest.permission.ACCESS_FINE_LOCATION
                )
            ) {
                // Show an explanation to the user *asynchronously* -- don't block
                // this thread waiting for the user's response! After the user
                // sees the explanation, try again to request the permission.
                AlertDialog.Builder(this)
                    .setTitle("Location Permission Needed")
                    .setMessage("This app needs the Location permission, please accept to use geographically bound reminders")
                    .setPositiveButton(
                        "OK"
                    ) { _, _ ->
                        //Prompt the user once explanation has been shown
                        requestLocationPermission()
                    }
                    .create()
                    .show()
            } else {
                // No explanation needed, we can request the permission.
                requestLocationPermission()
            }
        } else {
            checkBackgroundLocation()
        }
    }

    private fun checkBackgroundLocation() {
        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_BACKGROUND_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            val builder: AlertDialog.Builder = AlertDialog.Builder(this)
            builder.setTitle("Background Permission advised")
                .setMessage("This app runs gps in the background. If you don't locations all the time, your system might kill the service leading to no reminders. \nPlease select 'Allow all the time'.")
                .setPositiveButton("Ok") { _, _ ->
                    requestBackgroundLocationPermission()
                }
            builder.create().show()
        }
    }

    private fun requestLocationPermission() {
        ActivityCompat.requestPermissions(
            this,
            arrayOf(
                Manifest.permission.ACCESS_FINE_LOCATION,
            ),
            MY_PERMISSIONS_REQUEST_LOCATION
        )
    }

    private fun requestBackgroundLocationPermission() {
        ActivityCompat.requestPermissions(
            this,
            arrayOf(
                Manifest.permission.ACCESS_BACKGROUND_LOCATION
            ),
            MY_PERMISSIONS_REQUEST_BACKGROUND_LOCATION
        )
    }

    override fun onStart() {
        super.onStart()
        val context = this
        PreferenceManager.getDefaultSharedPreferences(context)
        //.registerOnSharedPreferenceChangeListener(context)

        val intentLocationService = Intent(this, LocationService::class.java)

        //Start listening to location updates
        bindService(
            intentLocationService, locationServiceConnection,
            BIND_AUTO_CREATE
        )
    }

    override fun onStop() {
        if (isServiceBound) {
            // Unbind from the service. This signals to the service that this activity is no longer
            // in the foreground, and the service can respond by promoting itself to a foreground
            // service.
            unbindService(locationServiceConnection)
            isServiceBound = false
        }
        val context = this
        PreferenceManager.getDefaultSharedPreferences(context)
        //.unregisterOnSharedPreferenceChangeListener(context)
        super.onStop()
    }

    /**
     * Handle navigation when the user chooses Up from the action bar.
     */
    override fun onSupportNavigateUp(): Boolean {
        return navController.navigateUp() || super.onSupportNavigateUp()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            MY_PERMISSIONS_REQUEST_LOCATION -> {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                    // permission was granted, yay! Do the
                    // location-related task you need to do.
                    if (ContextCompat.checkSelfPermission(
                            this,
                            Manifest.permission.ACCESS_FINE_LOCATION
                        ) == PackageManager.PERMISSION_GRANTED
                    ) {
                        locationService!!.requestLocationUpdates()

                        // Bind to the service. If the service is in foreground mode, this signals to the service
                        // that since this activity is in the foreground, the service can exit foreground mode.
                        bindService(
                            Intent(this, LocationService::class.java), locationServiceConnection,
                            BIND_AUTO_CREATE
                        )

                        // Now check background location
                        checkBackgroundLocation()
                    }

                } else {

                    // permission denied, boo! Disable the
                    // functionality that depends on this permission.
                    Toast.makeText(this, "How am I supposed to do something useful?", Toast.LENGTH_LONG)
                        .show()

                    // Check if we are in a state where the user has denied the permission and
                    // selected Don't ask again
                    if (!ActivityCompat.shouldShowRequestPermissionRationale(
                            this,
                            Manifest.permission.ACCESS_FINE_LOCATION
                        )
                    ) {
                        startActivity(
                            Intent(
                                Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                                Uri.fromParts("package", this.packageName, null),
                            ),
                        )
                    }
                }
                return
            }
            MY_PERMISSIONS_REQUEST_BACKGROUND_LOCATION -> {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                    // permission was granted, yay! Do the
                    // location-related task you need to do.
                    if (ContextCompat.checkSelfPermission(
                            this,
                            Manifest.permission.ACCESS_FINE_LOCATION
                        ) == PackageManager.PERMISSION_GRANTED
                    ) {
                        Toast.makeText(
                            this,
                            "All Location Permissions Granted.",
                            Toast.LENGTH_LONG
                        ).show()
                    }
                } else {

                    // permission denied, boo! Disable the
                    // functionality that depends on this permission.
                    Toast.makeText(this, "Your loss...", Toast.LENGTH_LONG)
                        .show()
                }
                return

            }
        }
    }

    companion object {
        private const val MY_PERMISSIONS_REQUEST_LOCATION = 99
        private const val MY_PERMISSIONS_REQUEST_BACKGROUND_LOCATION = 66
    }
}
